fairseq_cli/preprocess.py

```diff
@@ -45,6 +46,21 @@ def main(args):
 
     task = tasks.get_task(args.task)
 
+    def parse_source_args(args):
+        if args.srcdict:
+            if len(args.source_lang) != len(args.srcdict):
+                raise RuntimeError(
+                    "Number of source languages ({}) must match the number of source dictionaries ({}).".format(
+                        len(args.source_lang), len(args.srcdict)))
+            args.srcdict_uniq = list(set(args.srcdict))
+            dict_index = [args.srcdict_uniq.index(d) for d in args.srcdict]
+            return dict_index
+        else:
+            source_langs_stripped = [re.sub(r'\d+$', '', lang) for lang in args.source_lang]
+            args.srcdict_lang = list(set(source_langs_stripped))
```

Comment:
Similar to the solution above. Instead of loading, using the set of unique languages, I build each dictionary only once on line 125 and map the dictionaries to original "languages" again on line 129.